adam/adam_6224.py

```diff
@@ -0,0 +1,628 @@
+        if  (type_code == int("0182", 16) and not(value >= 0 and value <= 0.02)) or\
+                (type_code == int("0180", 16) and not(value >= 0.004 and value <= 0.02)) or\
+                (type_code == int("0148", 16) and not(value >= 0 and value <= 10))   or\
+                (type_code == int("0147", 16) and not(value >= 0 and value <= 5)) or\
+                (type_code == int("0143", 16) and not(value >= -10 and value <= 10)) or\
+                (type_code == int("0142", 16) and not(value >= -5 and value <= 5)):
+            if type == 0: tmp = self.analog_output_values[channel]
+            elif type == 1: tmp = self.analog_output_safety_values[channel]
+            else: tmp = self.analog_output_statup_values[channel]
+            self.error_stream('Illegal value')
+            raise ValueError
+        else:
+            if   type_code == int("0182", 16) : tmp = int(4095 * value / 0.02)
+            elif type_code == int("0180", 16) : tmp = int(4095 * (value - 0.004) / 0.016)
+            elif type_code == int("0148", 16) : tmp = int(4095 * value / 10)
+            elif type_code == int("0147", 16) : tmp = int(4095 * value/ 5 )
+            elif type_code == int("0143", 16) : tmp = int(4095 * (value + 10) / 20)
+            elif type_code == int("0142", 16) : tmp = int(4095 * (value + 5) / 10)
+            else:
+                if type == 0 :      tmp = self.analog_output_values[channel]
+                elif type == 1 :    tmp = self.analog_output_safety_values[channel]
+                else:               tmp = self.analog_output_statup_values[channel]
+
+        return tmp
+        # PROTECTED REGION END #    //  ADAM6224.decode_value
+
+    def encode_value(self, value, channel):
+        # PROTECTED REGION ID(ADAM6224.encode_value) ENABLED START #
+        type_code = self.analog_output_types[channel]
+        if   type_code == int("0182", 16): tmp = 0.02  * value/4095.0
+        elif type_code == int("0180", 16): tmp = (0.016 * value/4095.0) + 0.004
+        elif type_code == int("0148", 16): tmp = 10    * value/4095.0
+        elif type_code == int("0147", 16): tmp = 5     * value/4095.0
+        elif type_code == int("0143", 16): tmp = (20   * value/4095.0) - 10
+        elif type_code == int("0142", 16): tmp = (10   * value/4095.0) - 5
+        else: tmp = self.analog_output_values[channel]
+        return tmp
+        # PROTECTED REGION END #    //  ADAM6224.encode_value
+
+    def decode_type_code(self, channel):
+        # PROTECTED REGION ID(ADAM6224.decode_type_code) ENABLED START #
+        return self.type_code_dictionary[self.analog_output_types[channel]]
+        # PROTECTED REGION END #    //  ADAM6224.decode_type_code
+
+    def code_type_code(self, value="0-20mA"):
+        # PROTECTED REGION ID(ADAM6224.code_type_code) ENABLED START #
+        return self.type_code_dictionary[value]
+        # PROTECTED REGION END #    //  ADAM6224.code_type_code
+
+
+    # --------
+    # Commands
+    # --------
```

Comment:
I'd suggest to do all of the reading from ADAMs in a dedicated, polled command. It should update variables of the Tango Device. That way an operation of reading an attribute doesn't affect the controlled device.